streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -107,10 +107,10 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
             // we can skip flushing to downstream as well as writing to underlying store
             if (rawNewValue != null || rawOldValue != null) {
                 // we need to get the old values if needed, and then put to store, and then flush
-                wrapped().put(entry.key(), entry.newValue());
-
                 final ProcessorRecordContext current = context.recordContext();
                 context.setRecordContext(entry.entry().context());
+                wrapped().put(entry.key(), entry.newValue());
```

Comment:
Why reorder the steps here?

Reply:
This is actually important. We were not previously setting the record context before passing the cache-evicted record down to the lower store layers. Previously, the context was incorrectly not set during that operation, and if stores relied on the record context (via the old ProcessorContext), they were getting the wrong metadata.

Apparently, this work is the first time we added a feature in Streams that actually relied on that metadata. What is happening now is that we use that metadata to set the Position in the lower store, and if it's not set, then we get an error.

Reply:
I think there's some rationales to not set the processor context for the evicted record, e.g. let's say we first put record A offset 0 into the cache, and then record B offset 1, and then record C offset 2, the third insert caused the cache to evict A, which will then be written to the underlying store and also be forwarded downstreams, if we set the metadata as offset `2` for record `A` that may not be correct --- but I admit that if we do not set it, then it would be either some order context like offset `1` or even just a null, but it seems forwarding `A` with offset `2` is not appropriate still since we would prefer to forward `A` with offset `0` ideally

Reply:
I'm not sure I follow; we always did set the context, we just did it (incorrectly) between the `wrapped().put` and forwarding downstream. Why should we do the `wrapped().put` with the wrong context and then forward with the right context?

Taking your example sequence, the correct offset for record `A` is 0.
The old behavior was that we would do `wrapped().put(A)` with offset 2 and then forward `A` with offset 0.
The new behavior is that we do `wrapped().put(A)` with offset 0 and then forward `A` with offset 0.

There's no scenario in which we would forward A with offset 2.

Reply:
You are right! This is a good fix.